contrib/hashicorp/vault/vault_test.go

```diff
@@ -0,0 +1,281 @@
+	assert.Nil(client)
+	assert.Error(err)
+}
+
+const (
+	secretMountPath = "/ns1/ns2/secret"
+)
+
+func setupServer(t *testing.T) *httptest.Server {
+	storage := make(map[string]string)
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method == "PUT" {
+			fmt.Fprintln(w, "{}")
+			defer r.Body.Close()
+			bodyBytes, err := ioutil.ReadAll(r.Body)
+			body := string(bodyBytes)
+			if err != nil {
+				t.Error(err)
+				return
+			}
+			storage[r.URL.Path] = body
+		} else if r.Method == "GET" {
+			if value, ok := storage[r.URL.Path]; ok {
+				secret := vault.Secret{}
+				secret.Data = make(map[string]interface{})
+				json.Unmarshal([]byte(value), &secret.Data)
+				secretJson, err := json.Marshal(secret)
+				if err != nil {
+					t.Error(err)
+					return
+				}
+				fmt.Fprintf(w, "%s\n", secretJson)
+			} else {
+				http.Error(w, "No data for key.", http.StatusNotFound)
+			}
+		}
+	}))
+	return ts
+}
+
+func cleanupServer(ts *httptest.Server) {
+	ts.Close()
+}
+
+func setupVault(ts *httptest.Server) (*vault.Client, error) {
+	var config = &vault.Config{
+		HttpClient: NewHttpClient(),
+		Address:    ts.URL,
+	}
+
+	client, err := vault.NewClient(config)
+	client.SetToken("myroot")
+
+	secretMount := vault.MountInput{
+		Type:        "kv",
+		Description: "Test KV Store",
+		Local:       true,
+	}
+	err = client.Sys().Mount(secretMountPath, &secretMount)
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
+func cleanupVault(client *vault.Client) {
+	client.Sys().Unmount(secretMountPath)
+}
+
+func TestMountReadWrite(t *testing.T) {
+	assert := assert.New(t)
+	mt := mocktracer.Start()
+	defer mt.Stop()
+
+	ts := setupServer(t)
+	defer cleanupServer(ts)
+	client, err := setupVault(ts)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	defer cleanupVault(client)
+
+	key := secretMountPath + "/test"
+	secretData := map[string]interface{}{"Key1": "Val1", "Key2": "Val2"}
+	secret, err := client.Logical().Write(key, secretData)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	secret, err = client.Logical().Read(key)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	assert.Equal(secret.Data["Key1"], secretData["Key1"])
+	assert.Equal(secret.Data["Key2"], secretData["Key2"])
+
+	spans := mt.FinishedSpans()
+	assert.Len(spans, 3)
+	mount := spans[0]
+	write := spans[1]
+	read := spans[2]
+
+	fullPath := "/v1" + key
+
+	// Mount operation
+	assert.Equal("vault", mount.Tag(ext.ServiceName))
+	assert.Equal("/v1/sys/mounts/ns1/ns2/secret", mount.Tag(ext.HTTPURL))
+	assert.Equal(http.MethodPost, mount.Tag(ext.HTTPMethod))
+	assert.Equal(http.MethodPost+" /v1/sys/mounts/ns1/ns2/secret", mount.Tag(ext.ResourceName))
+	assert.Equal(ext.SpanTypeVault, mount.Tag(ext.SpanType))
+	assert.Equal(200, mount.Tag(ext.HTTPCode))
+	assert.Zero(mount.Tag(ext.Error))
+	assert.Zero(mount.Tag(ext.ErrorDetails))
+	assert.Zero(mount.Tag("vault.namespace"))
+
+	// Write key
```

Comment:
I think it would be much nicer, compact, and easier to follow if each one of these groups would be a separate test. Ideally as sub-tests of a larger encompassing test.